File: lib/features/stats/data/stats_repository.dart
import 'package:drift/drift.dart';

import '../../../data/db/app_database.dart';
import '../domain/stats_period.dart';

abstract interface class StatsRepository {
  Future<ListeningStats> getStats(StatsPeriod period);

  Stream<ListeningStats> watchStats(StatsPeriod period);

  Future<void> applyRetentionPolicy(int? retentionDays);

  Future<void> deleteAllHistory();
}

class StatsRepositoryImpl implements StatsRepository {
  const StatsRepositoryImpl({required AppDatabase database}) : _db = database;

  final AppDatabase _db;

  @override
  Future<ListeningStats> getStats(StatsPeriod period) async {
    final since = period.since;

    final sessions = await _sessionsSince(since);

    var totalSeconds = 0;
    var timeSavedSeconds = 0;
    final podcastMap = <int, ({String title, int seconds, int count})>{};

    for (final s in sessions) {
      totalSeconds += s.durationSeconds;

      // Time saved vs 1.0x: duration * (1 - 1/speed)
      if (s.speed > 1.0) {
        timeSavedSeconds += (s.durationSeconds * (1 - 1 / s.speed)).round();
      }

      final existing = podcastMap[s.podcastId];
      final title = await _podcastTitle(s.podcastId);
      podcastMap[s.podcastId] = (
        title: title,
        seconds: (existing?.seconds ?? 0) + s.durationSeconds,
        count: (existing?.count ?? 0) + 1,
      );
    }

    final episodesCompleted = await _episodesCompleted(since);

    final perPodcast =
        podcastMap.entries
            .map(
              (e) => PodcastStats(
                podcastId: e.key,
                podcastTitle: e.value.title,
                totalSeconds: e.value.seconds,
                episodeCount: e.value.count,
              ),
            )
            .toList()
          ..sort((a, b) => b.totalSeconds.compareTo(a.totalSeconds));

    return ListeningStats(
      totalSeconds: totalSeconds,
      timeSavedSeconds: timeSavedSeconds,
      episodesCompleted: episodesCompleted,
      perPodcast: perPodcast,
    );
  }

  @override
  Stream<ListeningStats> watchStats(StatsPeriod period) {
    final query = _db.select(_db.listeningSessions);
    if (period.since != null) {
      query.where((s) => s.date.isBiggerOrEqualValue(period.since!));
    }
    return query.watch().asyncMap((_) => getStats(period));
  }

  @override
  Future<void> applyRetentionPolicy(int? retentionDays) async {
    if (retentionDays == null) return;
    final cutoff = DateTime.now().toUtc().subtract(
      Duration(days: retentionDays),
    );
    await (_db.delete(
      _db.listeningSessions,
    )..where((s) => s.date.isSmallerThanValue(cutoff))).go();
  }

  @override
  Future<void> deleteAllHistory() async {
    await _db.delete(_db.listeningSessions).go();
  }

  Future<List<ListeningSessionRow>> _sessionsSince(DateTime? since) {
    final query = _db.select(_db.listeningSessions);
    if (since != null) {
      query.where((s) => s.date.isBiggerOrEqualValue(since));
    }
    return query.get();
  }

  Future<int> _episodesCompleted(DateTime? since) async {
    final query = _db.select(_db.episodes)
      ..where((e) => e.playedAt.isNotNull());
    if (since != null) {
      query.where((e) => e.playedAt.isBiggerOrEqualValue(since));
    }
    final rows = await query.get();
    return rows.length;
  }

  Future<String> _podcastTitle(int podcastId) async {
    final row = await (_db.select(
      _db.podcasts,
    )..where((p) => p.id.equals(podcastId))).getSingleOrNull();
    return row?.title ?? 'Unknown Podcast';
  }
}

File: EarshotSwift/Earshot/Features/Stats/Data/StatsRepository.swift
import Foundation
import SwiftData

/// Aggregates ``ListeningSession`` rows into ``ListeningStats``, applies the
/// history retention policy, deletes all history, and exports sessions as CSV.
/// Mirrors the Flutter `StatsRepositoryImpl`. Pure math lives in ``StatsLogic``.
@MainActor
final class StatsRepository {
    private let context: ModelContext

    init(context: ModelContext) {
        self.context = context
    }

    /// Computes stats for `period`. `includeStreak` gates the opt-in streak so
    /// it stays off (and uncomputed) by default.
    func stats(
        for period: StatsPeriod,
        now: Date = .now,
        includeStreak: Bool = false
    ) -> ListeningStats {
        let since = period.since(now: now)
        let sessions = sessionsSince(since)

        var totalSeconds = 0
        var timeSavedSeconds = 0
        var byPodcast: [String: (seconds: Int, count: Int)] = [:]

        for session in sessions {
            totalSeconds += session.durationSeconds
            timeSavedSeconds += StatsLogic.timeSavedBySpeed(
                durationSeconds: session.durationSeconds, speed: session.speed
            )
            let title = session.podcast?.title ?? "Unknown Podcast"
            let existing = byPodcast[title] ?? (0, 0)
            byPodcast[title] = (existing.seconds + session.durationSeconds, existing.count + 1)
        }

        let perPodcast = byPodcast
            .map { PodcastStat(podcastTitle: $0.key, totalSeconds: $0.value.seconds, episodeCount: $0.value.count) }
            .sorted { $0.totalSeconds > $1.totalSeconds }

        let streak = includeStreak
            ? StatsLogic.currentStreak(sessionDates: allSessionDates(), now: now)
            : 0

        return ListeningStats(
            totalSeconds: totalSeconds,
            timeSavedSeconds: timeSavedSeconds,
            episodesCompleted: episodesCompleted(since: since),
            currentStreakDays: streak,
            perPodcast: perPodcast
        )
    }

    /// Deletes sessions older than `days`. A non-positive value keeps everything.
    func applyRetention(days: Int, now: Date = .now) {
        guard days > 0 else { return }
        let cutoff = now.addingTimeInterval(-Double(days) * 86_400)
        let stale = (try? context.fetch(FetchDescriptor<ListeningSession>())) ?? []
        var removed = 0
        for session in stale where session.date < cutoff {
            context.delete(session)
            removed += 1
        }
        if removed > 0 {
            save()
            AppLog.data.info("Stats retention removed \(removed) old session(s)")
        }
    }

    func deleteAllHistory() {
        for session in (try? context.fetch(FetchDescriptor<ListeningSession>())) ?? [] {
            context.delete(session)
        }
        save()
        AppLog.data.info("Deleted all listening history")
    }

    /// All sessions as CSV (date,podcast,episode,seconds,speed), newest first.
    func csv(now: Date = .now) -> String {
        let formatter = ISO8601DateFormatter()
        let rows = sessionsSince(nil).sorted { $0.date > $1.date }
        var lines = ["date,podcast,episode,duration_seconds,speed"]
        for session in rows {
            let fields = [
                formatter.string(from: session.date),
                session.podcast?.title ?? "Unknown Podcast",
                session.episode?.title ?? "",
                String(session.durationSeconds),
                String(format: "%g", session.speed),
            ].map(Self.escapeCSV)
            lines.append(fields.joined(separator: ","))
        }
        return lines.joined(separator: "\n")
    }

    // MARK: Internals

    private func sessionsSince(_ since: Date?) -> [ListeningSession] {
        let all = (try? context.fetch(FetchDescriptor<ListeningSession>())) ?? []
        guard let since else { return all }
        return all.filter { $0.date >= since }
    }

    private func allSessionDates() -> [Date] {
        ((try? context.fetch(FetchDescriptor<ListeningSession>())) ?? []).map(\.date)
    }

    private func episodesCompleted(since: Date?) -> Int {
        let played = (try? context.fetch(FetchDescriptor<Episode>())) ?? []
        return played.filter { episode in
            guard let playedAt = episode.playedAt else { return false }
            guard let since else { return true }
            return playedAt >= since
        }.count
    }

    private static func escapeCSV(_ field: String) -> String {
        guard field.contains(",") || field.contains("\"") || field.contains("\n") else {
            return field
        }
        return "\"\(field.replacingOccurrences(of: "\"", with: "\"\""))\""
    }

    private func save() {
        guard context.hasChanges else { return }
        do {
            try context.save()
        } catch {
            AppLog.data.error("Stats save failed: \(error.localizedDescription, privacy: .public)")
        }
    }
}
